projects/packages/publicize/src/class-keyring-service-helper.php

```diff
@@ -0,0 +1,341 @@
+<?php
+/**
+ * Utilities to interact with a Keyring instance.
+ * Used for Publicize as well as the Site Verification tools.
+ *
+ * @package automattic/jetpack-publicize
+ */
+
+namespace Automattic\Jetpack\Publicize;
+
+use Automattic\Jetpack\Connection\Secrets;
+
+/**
+ * A series of utilities to interact with a Keyring instance.
+ */
+class Keyring_Service_Helper {
+	/**
+	 * Class instance
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @var Jetpack_Keyring_Service_Helper
+	 */
+	private static $instance = null;
+
+	/**
+	 * Whether the `sharing` page is registered.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @var bool
+	 */
+	private static $is_sharing_page_registered = false;
+
+	/**
+	 * Initialize instance.
+	 *
+	 * @since $$next-version$$
+	 */
+	public static function init() {
+		if ( self::$instance === null ) {
+			self::$instance = new Jetpack_Keyring_Service_Helper();
+		}
+
+		return self::$instance;
+	}
+
+	const SERVICES = array(
+		'facebook'                 => array(
+			'for' => 'publicize',
+		),
+		'twitter'                  => array(
+			'for' => 'publicize',
+		),
+		'linkedin'                 => array(
+			'for' => 'publicize',
+		),
+		'tumblr'                   => array(
+			'for' => 'publicize',
+		),
+		'path'                     => array(
+			'for' => 'publicize',
+		),
+		'google_plus'              => array(
+			'for' => 'publicize',
+		),
+		'google_site_verification' => array(
+			'for' => 'other',
+		),
+	);
+
+	/**
+	 * Constructor
+	 *
+	 * @since $$next-version$$
+	 */
+	private function __construct() {
+		add_action( 'admin_menu', array( __CLASS__, 'register_sharing_page' ) );
+
+		add_action( 'load-settings_page_sharing', array( __CLASS__, 'admin_page_load' ), 9 );
+	}
+
+	/**
+	 * We need a `sharing` page to be able to connect and disconnect services.
+	 *
+	 * @since $$next-version$$
+	 */
+	public static function register_sharing_page() {
+		if ( self::$is_sharing_page_registered ) {
+			return;
+		}
+
+		self::$is_sharing_page_registered = true;
```

Comment:
What happens if the sharing menu is already registered by a different plugin, like Jetpack? Should we take that into account?

Reply:
For now, I pushed a commit that goes up a level, to the constructor, and comments out the hook that calls this function. The Reach team can figure out how best to handle this.